public/app/core/components/Page/EditableTitle.tsx

```diff
@@ -0,0 +1,117 @@
+import { css } from '@emotion/css';
+import React, { useCallback, useState } from 'react';
+
+import { GrafanaTheme2 } from '@grafana/data';
+import { isFetchError } from '@grafana/runtime';
+import { Field, IconButton, Input, useStyles2 } from '@grafana/ui';
+import { H1 } from '@grafana/ui/src/unstable';
+
+export interface Props {
+  value: string;
+  onEdit: (newValue: string) => Promise<void>;
+}
+
+export const EditableTitle = ({ value, onEdit }: Props) => {
+  const styles = useStyles2(getStyles);
+  const [localValue, setLocalValue] = useState<string>();
+  const [isEditing, setIsEditing] = useState(false);
+  const [isEditInProgress, setIsEditInProgress] = useState(false);
+  const [errorMessage, setErrorMessage] = useState<string>();
+
+  const onCommitChange = useCallback(
+    async (event: React.FormEvent<HTMLInputElement>) => {
+      const newValue = event.currentTarget.value;
+
+      if (!newValue) {
+        setErrorMessage('Please enter a title');
+      } else if (newValue === value) {
+        // no need to bother saving if the value hasn't changed
+        // just clear any previous error messages and exit edit mode
+        setErrorMessage(undefined);
+        setIsEditing(false);
+      } else {
+        setIsEditInProgress(true);
+        try {
+          await onEdit(newValue);
+          setErrorMessage(undefined);
+          setIsEditing(false);
+        } catch (error) {
+          if (isFetchError(error)) {
+            setErrorMessage(error.data.message);
+          } else if (error instanceof Error) {
+            setErrorMessage(error.message);
+          }
+        }
+        setIsEditInProgress(false);
+      }
+    },
+    [onEdit, value]
+  );
+
+  return !isEditing ? (
+    <div className={styles.textContainer}>
+      <div className={styles.textWrapper}>
+        {/*
+          use localValue if it exists
+          this is to prevent the title from flickering back to the old value after the user has edited
+          caused by the delay between the save completing and the new value being refetched
+        */}
+        <H1 truncate>{localValue ?? value}</H1>
```

Comment:
I think there's an iffy bit here where the parent cannot change the `value` prop after its ever been edited?

I _think_ this is an instance where we want to react to changes in the `value` prop and sync it to state.

```ts
const [localValue, setLocalValue] = useState(value);
useEffect(() => {
  setLocalValue(value);
}, [value])
```

(this should not cause a re-render on first mount as the value doesn't change)